File: apps/desktop/src/lib/ai.test.ts
import { describe, it, expect } from 'vitest'
import { canTransitionAi, validateAiSummarizePayload, validateAiTranscribePayload, validateAiResultPayload, shouldRateLimit, localSummarize } from './ai'

describe('ai validation', () => {
  it('state transitions', () => {
    expect(canTransitionAi('IDLE','QUEUED')).toBe(true)
    expect(canTransitionAi('QUEUED','LOCAL')).toBe(true)
    expect(canTransitionAi('QUEUED','CLOUD')).toBe(true)
    expect(canTransitionAi('LOCAL','DONE')).toBe(true)
    expect(canTransitionAi('LOCAL','CLOUD')).toBe(true)
    expect(canTransitionAi('CLOUD','DONE')).toBe(true)
    expect(canTransitionAi('IDLE','DONE')).toBe(false)
    expect(canTransitionAi('DONE','QUEUED')).toBe(false)
    expect(canTransitionAi('DONE','IDLE')).toBe(true)
  })
  it('validate summarize ok', () => {
    expect(validateAiSummarizePayload({notifications:[{app:'WhatsApp', body:'hello'}], maxLen:200})).toBe(null)
  })
  it('validate summarize empty', () => {
    expect(validateAiSummarizePayload({notifications:[], maxLen:200})).not.toBe(null)
  })
  it('validate summarize too many', () => {
    const many = Array(21).fill({app:'A', body:'hi'})
    expect(validateAiSummarizePayload({notifications:many})).not.toBe(null)
  })
  it('validate summarize body too long', () => {
    expect(validateAiSummarizePayload({notifications:[{app:'A', body:'a'.repeat(501)}]})).not.toBe(null)
  })
  it('validate transcribe ok', () => {
    const b64 = btoa(String.fromCharCode(...new Uint8Array(100).fill(0x42)))
    expect(validateAiTranscribePayload({audio_b64:b64, format:'opus', lang:'en'})).toBe(null)
    expect(validateAiTranscribePayload({audio_b64:b64, format:'wav'})).toBe(null)
  })
  it('validate transcribe invalid format', () => {
    const b64 = btoa('hi')
    expect(validateAiTranscribePayload({audio_b64:b64, format:'evil'})).not.toBe(null)
  })
  it('validate transcribe empty', () => {
    expect(validateAiTranscribePayload({audio_b64:'', format:'opus'})).not.toBe(null)
  })
  it('validate result ok', () => {
    expect(validateAiResultPayload({kind:'summarize', text:'hi', model:'llama.cpp-local'})).toBe(null)
    expect(validateAiResultPayload({kind:'transcribe', text:'hi', model:'whisper.cpp-local'})).toBe(null)
  })
  it('validate result invalid kind', () => {
    expect(validateAiResultPayload({kind:'evil', text:'hi', model:'m'})).not.toBe(null)
  })
  it('rate limit', () => {
    const ts: number[] = []
    for(let i=0;i<10;i++) expect(shouldRateLimit(ts, 1000)).toBe(false)
    expect(shouldRateLimit(ts, 1000)).toBe(true)
    expect(shouldRateLimit(ts, 70000)).toBe(false)
  })
  it('local summarize', () => {
    const notifs = [{app:'WhatsApp', body:'hello'}, {app:'WhatsApp', body:'hi'}, {app:'Gmail', body:'test'}]
    const s = localSummarize(notifs, 200)
    expect(s).toContain('3 notifications')
    expect(s).toContain('WhatsApp')
    expect(s.length).toBeLessThanOrEqual(200)
  })
})

File: crates/bridge-core/tests/ai_test.rs
use bridge_core::{AiState, validate_ai_summarize_payload, validate_ai_transcribe_payload, validate_ai_result_payload, should_rate_limit_ai};
use serde_json::json;
use base64::{engine::general_purpose::STANDARD as B64, Engine as _};

fn b64_audio(n: usize) -> String { B64.encode(vec![0x42u8; n]) }

#[test]
fn ai_state_valid() {
    assert!(AiState::Idle.can_transition(&AiState::Queued));
    assert!(AiState::Queued.can_transition(&AiState::Local));
    assert!(AiState::Queued.can_transition(&AiState::Cloud));
    assert!(AiState::Local.can_transition(&AiState::Done));
    assert!(AiState::Local.can_transition(&AiState::Cloud));
    assert!(AiState::Cloud.can_transition(&AiState::Done));
    assert!(AiState::Done.can_transition(&AiState::Idle));
}
#[test]
fn ai_state_invalid() {
    assert!(!AiState::Idle.can_transition(&AiState::Done));
    assert!(!AiState::Done.can_transition(&AiState::Queued));
}
#[test]
fn validate_summarize_ok() {
    let p = json!({"notifications":[{"app":"WhatsApp","body":"hello"}],"maxLen":200,"requestId":"req1"});
    assert!(validate_ai_summarize_payload(&p).is_ok());
}
#[test]
fn validate_summarize_empty() {
    let p = json!({"notifications":[],"maxLen":200});
    assert!(validate_ai_summarize_payload(&p).is_err());
}
#[test]
fn validate_summarize_too_many() {
    let many: Vec<_> = (0..21).map(|_| json!({"app":"A","body":"hi"})).collect();
    let p = json!({"notifications": many});
    assert!(validate_ai_summarize_payload(&p).is_err());
}
#[test]
fn validate_summarize_body_too_long() {
    let p = json!({"notifications":[{"app":"A","body": "a".repeat(501)}]});
    assert!(validate_ai_summarize_payload(&p).is_err());
}
#[test]
fn validate_transcribe_ok() {
    let b64 = b64_audio(100);
    let p = json!({"audio_b64":b64,"format":"opus","lang":"en","requestId":"req1"});
    assert!(validate_ai_transcribe_payload(&p).is_ok());
}
#[test]
fn validate_transcribe_invalid_format() {
    let b64 = b64_audio(10);
    let p = json!({"audio_b64":b64,"format":"evil"});
    assert!(validate_ai_transcribe_payload(&p).is_err());
}
#[test]
fn validate_transcribe_empty() {
    let p = json!({"audio_b64":"","format":"opus"});
    assert!(validate_ai_transcribe_payload(&p).is_err());
}
#[test]
fn validate_result_ok() {
    let p = json!({"kind":"summarize","text":"hi","model":"llama.cpp-local"});
    assert!(validate_ai_result_payload(&p).is_ok());
}
#[test]
fn validate_result_invalid_kind() {
    let p = json!({"kind":"evil","text":"hi","model":"m"});
    assert!(validate_ai_result_payload(&p).is_err());
}
#[test]
fn rate_limit_ai() {
    let mut v: Vec<i64> = Vec::new();
    for _ in 0..10 { assert!(!should_rate_limit_ai(&mut v, 1000, 10, 60000)); }
    assert!(should_rate_limit_ai(&mut v, 1000, 10, 60000));
    assert!(!should_rate_limit_ai(&mut v, 70000, 10, 60000));
}
#[test]
fn message_type_serde_ai() {
    let m = bridge_core::BridgeMessage::new(bridge_core::MessageType::AiSummarize, json!({"notifications":[{"app":"A","body":"hi"}]}));
    assert!(m.to_json().contains("ai.summarize"));
    let m2 = bridge_core::BridgeMessage::new(bridge_core::MessageType::AiTranscribe, json!({"audio_b64":b64_audio(10),"format":"opus"}));
    assert!(m2.to_json().contains("ai.transcribe"));
    let m3 = bridge_core::BridgeMessage::new(bridge_core::MessageType::AiResult, json!({"kind":"summarize","text":"hi","model":"x"}));
    assert!(m3.to_json().contains("ai.result"));
    let m4 = bridge_core::BridgeMessage::new(bridge_core::MessageType::PluginList, json!({}));
    assert!(m4.to_json().contains("plugin.list"));
}
